src/connmgr/connection.rs

```diff
@@ -1413,18 +1413,21 @@ async fn server_req_read_body<R: AsyncRead, W: AsyncWrite>(
 
     loop {
         match req_body.try_recv(body_buf.write_buf())? {
-            RecvStatus::Complete((), written) => {
-                body_buf.write_commit(written);
+            RecvStatus::Complete((), size) => {
+                body_buf.write_commit(size);
                 break;
             }
-            RecvStatus::Read((), written) => {
-                body_buf.write_commit(written);
+            RecvStatus::Read((), size) => {
+                body_buf.write_commit(size);
 
-                if written == 0 {
-                    // ABR: discard_while
-                    discard_while(zreceiver, pin!(req_body.add_to_buffer())).await?;
+                if size == 0 {
+                    return Err(Error::BufferExceeded);
                 }
             }
+            RecvStatus::NeedBytes(()) => {
+                // ABR: discard_while
+                discard_while(zreceiver, pin!(req_body.add_to_buffer())).await?;
+            }
         }
     }
```

Comment:
This loop reads a request body from a client into a fixed-sized buffer. Previously, it would try to read more bytes into the input buffer whenever `try_recv` returned zero, which could happen whenever bytes were needed or if the output buffer was full. Now, it only tries to read more bytes into the input buffer whenever `try_recv` returns `NeedBytes`.

This means the only change in behavior is what it does when the output buffer is full. Previously, it would try to read more bytes into the input buffer, either hanging the connection or eventually returning `BufferExceeded` if the input buffer filled up also. Now, it immediately returns `BufferExceeded`. So there should be no performance impact during regular operation, just an earlier error when the output is full and avoiding a connection hang.